### PR TITLE
Add execute driver

### PR DESCRIPTION
**Pull request by KazuCocoa**
Waiting for https://github.com/appium/appium-base-driver/pull/328

Added _execute_driver_ as the new feature.

You can try the feature as `$ pytest test/functional/ios/execute_driver_tests.py` after applying the base driver PR on your local. I'll update this PR if something changes in the base driver PR.

### CODE REVIEW

File: appium/webdriver/extensions/execute_driver.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+# limitations under the License.
+
+from selenium import webdriver
+
+from ..mobilecommand import MobileCommand as Command
+
+
+class ExecuteDriver(webdriver.Remote):
+
+    def execute_driver(self, script, script_type='webdriverio', timeout=None):
```
Comment (mykola-mokhnach) [inline appium/webdriver/extensions/execute_driver.py]: timeout_ms might be a better arg name

File: test/functional/ios/execute_driver_tests.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+class ExecuteDriverTests(unittest.TestCase):
+    def setUp(self):
+        desired_caps = desired_capabilities.get_desired_capabilities('UICatalog.app.zip')
+        self.driver = webdriver.Remote('http://localhost:4723/wd/hub', desired_caps)
+
+    def tearDown(self):
+        self.driver.quit()
+
+    def test_batch(self):
+        script = """
```
Comment (mykola-mokhnach) [inline test/functional/ios/execute_driver_tests.py]: use https://docs.python.org/2/library/textwrap.html#textwrap.dedent

File: appium/webdriver/extensions/execute_driver.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+            script_type (string): The name of the script type. Defaults to 'webdriverio'.
+            timeout (optional): The number of `ms` Appium should wait for the script to finish before killing it due to timeout.
+
+        Usage:
+            self.driver.execute_driver(script='return [];')
+            self.driver.execute_driver(script='return [];', script_type='webdriverio')
+            self.driver.execute_driver(script='return [];', script_type='webdriverio', timeout=10000)
+
+        Return:
+            Dir[str, any]: The result of the script. It has 'result' and 'logs' keys.
```
Comment (ki4070ma) [inline appium/webdriver/extensions/execute_driver.py]: Dir -> Dict?

File: appium/webdriver/extensions/execute_driver.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+            script (string): The string consisting of the script itself
+            script_type (string): The name of the script type. Defaults to 'webdriverio'.
+            timeout (optional): The number of `ms` Appium should wait for the script to finish before killing it due to timeout.
+
+        Usage:
+            self.driver.execute_driver(script='return [];')
+            self.driver.execute_driver(script='return [];', script_type='webdriverio')
+            self.driver.execute_driver(script='return [];', script_type='webdriverio', timeout=10000)
+
+        Return:
```
Comment (ki4070ma) [inline appium/webdriver/extensions/execute_driver.py]: Return -> Returns

https://github.com/google/styleguide/blob/gh-pages/pyguide.md#38-comments-and-docstrings

File: test/unit/webdriver/device/execute_driver_test.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+        """
+        result = driver.execute_driver(script=script)
+        # Python client convert an element item as WebElement in the result
+        assert result['result'][0].id == '39000000-0000-0000-D39A-000000000000'
+        assert result['result'][1]['y'] == 237
+
+        d = get_httpretty_request_body(httpretty.last_request())
+        assert d['script'] == script
+        assert d['type'] == 'webdriverio'
+        assert 'timeout' not in d
```
Comment (ki4070ma) [inline test/unit/webdriver/device/execute_driver_test.py]: ```'timeout' not in d.keys()``` ?

Comment (KazuCocoa) [reply test/unit/webdriver/device/execute_driver_test.py]: Current way is fine 
```
>>> d = {'timeout': 1}
>>> 'timeout' in d
True
>>> 'timeout' in d.keys()
True
```

File: appium/webdriver/extensions/execute_driver.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+            def __init__(self, response):
+                self.result = response['result']
+                self.logs = response['logs']
+
+        option = {'script': script, 'type': script_type}
+        if timeout_ms is not None:
+            option['timeout'] = timeout_ms
+
+        response = self.execute(Command.EXECUTE_DRIVER, option)['value']
+        return Result(response)
```
Comment (ki4070ma) [inline appium/webdriver/extensions/execute_driver.py]: Below codes can be available.

```
return response['result'], response['logs']
```

File: test/functional/ios/execute_driver_tests.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    def test_batch_combination_python_script(self):
+        script = """
+            console.warn('warning message');
+            const element = await driver.findElement('accessibility id', 'Buttons');
+            const rect = await driver.getElementRect(element.ELEMENT);
+            return [element, rect];
+        """
+
+        result = self.driver.execute_driver(script=textwrap.dedent(script))
+        r = result.result[0].rect
```
Comment (mykola-mokhnach) [inline test/functional/ios/execute_driver_tests.py]: I's say itis a bit confusing to have `result.result` mapping. What if we rename the retuned variable to `response` ?

Comment (mykola-mokhnach) [reply test/functional/ios/execute_driver_tests.py]: same below

File: appium/webdriver/extensions/execute_driver.py (hunk 6 above)
Comment (KazuCocoa) [reply appium/webdriver/extensions/execute_driver.py]: Yes, but I hope users do not consider the order, [0] is _result_ and [1] is _logs_
thanks

File: test/functional/ios/execute_driver_tests.py (hunk 7 above)
Comment (KazuCocoa) [reply test/functional/ios/execute_driver_tests.py]: ah. Nice.  Yeah, I forgot re-naming it

**Pull request merged**